File: go/types/number_util.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package types

import "math"

const maxSafeInteger = float64(9007199254740991) // 2 ** 53 -1

func float64IsInt(f float64) bool {
	return math.Trunc(f) == f
}

// convert float64 to int64 where f == i / 2^exp
func float64ToIntExp(f float64) (int64, int) {
	if f == 0 {
		return 0, 0
	}

	isNegative := math.Signbit(f)
	f = math.Abs(f)

	exp := 0
	// Really large float, bring down to max safe integer so that it can be correctly represented by float64.
	for f > maxSafeInteger {
		f /= 2
		exp--
	}

	for !float64IsInt(f) {
		f *= 2
		exp++
	}
	if isNegative {
		f *= -1
	}
	return int64(f), exp
}

// returns float value == i / 2^exp
func intExpToFloat64(i int64, exp int) float64 {
	if exp == 0 {
		return float64(i)
	}

	return float64(i) / math.Pow(2, float64(exp))
}

File: js/noms/src/number-util.js
// @flow

// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

function float64IsInt(f: number): boolean {
  return Math.trunc(f) === f;
}

/**
 * Converts a number to [int, exp] where f = i / 2^exp
 */
export function floatToIntExp(f: number): [number, number] {
  if (f === 0) {
    return [0, 0];
  }

  const sign = Math.sign(f);
  f = Math.abs(f);

  let exp = 0;

  // Really large float, bring down to max safe integer so that it can be correctly represented by
  // float64.
  while (f > Number.MAX_SAFE_INTEGER) {
    f /= 2;
    exp--;
  }

  while (!float64IsInt(f)) {
    f *= 2;
    exp++;
  }
  return [sign * f, exp];
}

/**
 * Computes the float value of i / 2^exp
 */
export function intExpToFloat(i: number, exp: number): number {
  if (exp === 0) {
    return i;
  }
  return i / Math.pow(2, exp);
}
